### DISCUSSION TITLE
"requiring" monaco-editor in a "cleaner manner"

### DISCUSSION THREAD

**Discussion by frotaur** [microsoft/monaco-editor#3826] [Q&A]
### Context

- [X] This issue is not a bug report. *(please use a different template for reporting a bug)*
- [X] This issue is not a duplicate of an existing issue. *(please use the [search](https://github.com/microsoft/monaco-editor/issues) to find existing issues)*

### Description

I am having extreme trouble trying to integrate monaco-editor in an electron app. Please note I am new to web-app development, so I am very sorry if my question is trivial/unfeasible, although I have been trying for two days now to come up with a satisfactory solution.

Currently, to load monaco-editor, I can do one of those two things :

- Use "AMDloader". In that case, I include "`<script src="./node_modules/monaco-editor/min/vs/loader.js"></script>`" in my index.html, and to access the monaco API I must wrap my calls with "`require(['vs/editor/editor.main'], <call-back function which has access to the monaco API>)`". This is acceptable, but will become very ugly, very quick, if I need to access the monaco API often. I will have to include this repeated "require" code everywhere.  
  There is actually a mild workaround, which is to define the monaco editor in an initial "require" statement, and keep a reference to the editor, so that I can access it outside. This is great, with one small caveat : the call-back function inside the require statement is executed asynchronously, so I need to give a slight initial delay for the call-back to resolve, otherwise I would be trying to access an undefined "editor" outside. Is there a better way to ensure that the "require" block has been executed ?

- Use "webpacker". In that case, I can just add import * as monaco from "monaco-editor" in my renderer.js file, and use the monaco API everywhere (which is very nice). However it comes with the very cumbersome fact that I must use webpacker (which would be need only for this npm package, everything else I can do natively), which requires me to do a long build each time I want to launch my program, and additionally, the output generated in "dist" is completely unreadable (which I guess is part of the point, but still).

So my question/request is : is there a way to get the best of both worlds ? How could I make it so that I can have access to the monaco API easily in my renderer.js, without having to introduce something like webpack ? For example, the "AMDLoader" solution would be perfect, if I didn't have to wrap everything into a require. I don't know react, but it seems there is a wrapper that allows you to use the monaco API easily in that framework. I don't think there is something equivalent for electron, but how hard would it be to develop something like that ?

N.B. : I have also tried emulating the "browser" samples, by simply including all the scripts in the html, but that doesn't mix well with electron, as it throws "Refused to execute inline script because it violates the following Content Security Policy..."

**Comment by montella1507** [comment]
Well... yes.

And you cannot import ANYTHING from 'monaco-editor' without Bundling EVERYTHING to main bundle. Liberary seems not to be very treeshakeable..

https://github.com/microsoft/monaco-editor/issues/3518

**Comment by Prinzhorn** [comment]
You are mixing a lot of things and I don't think monaco has anything to do with your problem. Bundling is standard nowadays and you will sooner or later run into the same issues with any library. I personally wouldn't recommend webpack, there are dozens of tools out there in different levels of abstractions (esbuild, Parcel, Rollup, Vite,...). Pick something.

>  I don't know react, but it seems there is a wrapper that allows you to use the monaco API easily in that framework. I don't think there is something equivalent for electron, but how hard would it be to develop something like that ?

That doesn't make sense as React and Electron concern two entirely unrelated things. Nothing keeps you from using React inside Electron. Hence there never will be a "Monaco for Electron" as this doesn't make sense. Like asking for a car with strawberry flavor.

> I have also tried emulating the "browser" samples, by simply including all the scripts in the html, but that doesn't mix well with electron, as it throws "Refused to execute inline script because it violates the following Content Security Policy..."

This has nothing to do with Electron. The Electron quick start just happens to have an inline CSP in their index.html. If you want to go that route nothing keeps you from using these examples without a bundler.

**Comment by frotaur** [comment]
I see, thanks for the precision. As I was expecting, there is a lot I don't understand yet, and this is valuable info. 

My "request", came simply from the naive understanding that I have of frontend and ES6. When I make several .js modules to use in my frontend, I simply use the "export" "import" keywords such that I can have access to stuff defined in the other file. I was just wondering if it was possible to implement something similar for monaco, where I would simply do the import statement, without having to "webpack" the whole application. My understanding (probably still incorrect) is that monaco does not use only .js files, but also .css and as such that is not doable ?

About the react comment, indeed, I have no knowledge of this library, so your comments are correct. When I was saying "for electron", it is simply the fact that in the samples there are several method shown to use monaco-editor in different cases, several of which are "for electron". But yet, the phrasing was not great.

As for the CSP, indeed you are correct, I will read about that and see if I can safely bypass it.
EDIT : Actually removing the CSP line in the HTML file allows me to use monaco-editor exactly as I wanted, THANK YOU for that remark. I will read about CSP and see how I can still include it, while running the monaco-editor scripts.

Thanks for the info, and sorry about my misconceptions !

**Comment by Prinzhorn** [comment]
> My "request", came simply from the naive understanding that I have of frontend and ES6. When I make several .js modules to use in my frontend, I simply use the "export" "import" keywords such that I can have access to stuff defined in the other file. I was just wondering if it was possible to implement something similar for monaco, where I would simply do the import statement, without having to "webpack" the whole application. My understanding (probably still incorrect) is that monaco does not use only .js files, but also .css and as such that is not doable ?

https://github.com/microsoft/monaco-editor/search?q=native+esm&type=issues

https://github.com/microsoft/monaco-editor/issues/886
https://github.com/microsoft/monaco-editor/issues/2605
https://github.com/microsoft/monaco-editor/issues/2795

**Comment by jogibear9988** [comment]
I use monaco npm package in a frontend without any bundeling. But cause of the loader in a very hackish way. I home they will sometime switch to an ES6 module.
They css loading could the be done by normal import. But as the import assertions proposal now got downgraded to stage 2 from 3 maybe this will still take a while...
